### ISSUE TITLE
Warn when creating `Set<void>`

### ISSUE THREAD

**Issue by jamesderlin** [dart-lang/sdk#58456]
Creating collection literals from `void` surprisingly doesn't generate any runtime errors nor analysis warnings:

```dart
void voidFunction() {}

void main() {
  var list = [voidFunction()];
  var set = {voidFunction()};
  var f = () => {voidFunction()};

  print(list); // Prints: [null]
  print(set); // Prints: {null}
  print(f()); // Prints: {null}
}
```
I'm confused why since my understanding was that `void` values are not allowed to be read.

I'm not sure if this somehow is intentionally not an error, but at a minimum, I think that we could warn if creating `List<void>` or `Set<void>`.  Those types don't seem like they'd ever be useful.  `Set<void>` is particularly bad since I've observed at least one person using `(arg) => { doSomething() }` when they meant either `(arg) => doSomething()` or `(arg) { doSomething(); }`. (Such a lint wouldn't help if `doSomething()` returned a non-`void` value, of course ,but I don't know what could be done about that other than requiring `=> { ... }` to be `=> <T>{ ... }` if a `Set` is actually desired.)

I tried this with version 2.13.4 of the Dart SDK.

**Comment by lrhn** [comment]
I don't think collection elements are on the allow-list for `void`-typed expressions.
(Not the version in the language specification. I don't think it was updated for Null Safety.)
If so, it should just be an error.

The list is (paraphrased):
* Expression-statement expression, `here;`.
* Initializer and increment of `for`, `for(here;...;here)...`.
* Cast, `here as SomeType`.
* Parenthesized expression `(here)`.
* Conditional expression branches, `...? here : here`.
* Null coalescing second operand, `... ?? here`
* Await, `await here`
* Returns (depending on function), `return here;`
* Arrow returms (depending on function), `=> here;`.
* Variable initializer, `var id = here;`.
* Actual argument for `void`-typed parameter, `<void>[].add(here)`, includes operators `[]` and `[]=`.
* Setter arguments, `<void>[1].first = here;`   
* Local variable assignment, `void x; x = here;`.
* Cascade assignment, `e..selectors.setter = here`. (I don't know why this isn't just a setter.)

Nothing mentions collection elements. I don't remember any feature which would have changed this (IIRC control-flow collections didn't, and Null Safety only affected returns).

(Now, if you desugar collection elements into `.add` calls, then you could think it falls under the "argument to `void` parameter". Desugaring shouldn't change behavior, and errors should be specified in terms of the code the user writes.)

**Comment by eernstg** [comment]
> Nothing mentions collection elements

Good catch! I considered `<void>[print('')]` to be OK because the expression `print('')` has type `void` and it is similar to an actual argument to a formal parameter whose declared type is `void`. However, we don't have a rule that matches this case precisely.

This means that implementations should report expressions of type `void` given as elements of list or set literals, or as keys or values of map elements, even in the case where the relevant type argument of the collection is `void`.

```dart
void v = null;

void main() {
  <void>[v]; // Error.
  <void>{v}; // Error.
  <void, void>{v: v}; // Two errors.
}
```

This would be a breaking change, however, which means that it should be processed as such. It probably doesn't break much code in practice. We may decide that the errors should be reported, or we might end up saying that it's better to adjust the language specification to allow it.

**Comment by eernstg** [comment]
I just created https://github.com/dart-lang/language/issues/4574 in order to clarify this matter at the language level. I'm blocking this issue on https://github.com/dart-lang/language/issues/4574.
